src/Engine/Runner.php

```diff
@@ -103,15 +104,30 @@ public function run($mode = self::MODE_CONTINUE)
 				$this->printer->printReset();
 			}
 
-			$this->driver->createTable();
-			$migrations = $this->driver->getAllMigrations();
-			$files = $this->finder->find($this->groups, array_keys($this->extensionsHandlers));
-			$toExecute = $this->orderResolver->resolve($migrations, $this->groups, $files, $mode);
-			$this->printer->printToExecute($toExecute);
-
-			foreach ($toExecute as $file) {
-				$queriesCount = $this->execute($file);
-				$this->printer->printExecute($file, $queriesCount);
+			$this->driver->beginTransaction();
+			try {
+				$this->driver->createTable();
+				$migrations = $this->driver->getAllMigrations();
+				$files = $this->finder->find($this->groups, array_keys($this->extensionsHandlers));
+				$toExecute = $this->orderResolver->resolve($migrations, $this->groups, $files, $mode);
+				$this->printer->printToExecute($toExecute);
+
+				foreach ($toExecute as $file) {
+					$queriesCount = $this->execute($file);
+					$this->printer->printExecute($file, $queriesCount);
+				}
+				$this->driver->commitTransaction();
+
+			} catch (\Exception $e) {
+				if ($mode === self::MODE_CONTINUE_FULL_ROLLBACK) {
+					// rollback all migrations executed in this run
+					$this->driver->rollbackTransaction();
+
+				} else if ($mode === self::MODE_CONTINUE) {
+					// commit migrations not including the failing one
+					$this->driver->commitTransaction();
```

Comment:
This is probably wrong. It migrations contains more sqls and only the last one is wrong (throws exception), all previous queries are commited, aren't they?